chain/cosmos/chain_node.go

```diff
@@ -745,14 +774,49 @@ func (tn *ChainNode) UnsafeResetAll(ctx context.Context) error {
 	return err
 }
 
-func (tn *ChainNode) CreatePool(ctx context.Context, keyName string, contractAddress string, swapFee float64, exitFee float64, assets []ibc.WalletAmount) error {
-	// TODO generate --pool-file
-	poolFilePath := "TODO"
-	_, err := tn.ExecTx(ctx,
+func (tn *ChainNode) CreatePool(ctx context.Context, keyName string, params ibc.PoolParams) (string, error) {
+	poolbz, err := json.Marshal(params)
+	if err != nil {
+		return "", err
+	}
+
+	poolFile := "pool.json"
+
+	fw := dockerutil.NewFileWriter(tn.logger(), tn.DockerClient, tn.TestName)
+	if err := fw.WriteFile(ctx, tn.VolumeName, poolFile, poolbz); err != nil {
+		return "", fmt.Errorf("failed to write pool file: %w", err)
+	}
+
+	if _, err := tn.ExecTx(ctx, keyName,
 		"gamm", "create-pool",
-		"--pool-file", poolFilePath,
+		"--pool-file", filepath.Join(tn.HomeDir(), poolFile),
+	); err != nil {
+		return "", fmt.Errorf("failed to create pool: %w", err)
+	}
+
+	stdout, _, err := tn.ExecQuery(ctx, "gamm", "num-pools")
+	if err != nil {
+		return "", fmt.Errorf("failed to query num pools: %w", err)
+	}
+	var res map[string]string
+	if err := json.Unmarshal(stdout, &res); err != nil {
+		return "", fmt.Errorf("failed to unmarshal query response: %w", err)
+	}
+
+	numPools, ok := res["num_pools"]
+	if !ok {
+		return "", fmt.Errorf("could not find number of pools in query response: %w", err)
+	}
+	return numPools, nil
+}
+
+func (tn *ChainNode) SwapExactAmountIn(ctx context.Context, keyName string, coinIn string, minAmountOut string, poolIDs []string, swapDenoms []string) (string, error) {
```

Comment:
Doc comment would also be useful here. As an outsider, the function name is not descriptive enough. 

Additionally, are `SwapExactAmountIn` and `CreatePool` osmosis specific? Do these methods apply to all chains? 

If not, one option is a new `cosmos.OsmosisChain` type. You can embed a `cosmos.ChainNode` to get all of its functionality. Otherwise, it appears as if all cosmos chains has this Osmosis-specific functionality.

Reply:
Or, another alternative is to separate data from function. In other words, instead of creating new methods off the type, create functions that take the chain as an argument. 

E.g. 

```go
// Ideally, *ChainNode is replaced by an interface. 
// https://bryanftan.medium.com/accept-interfaces-return-structs-in-go-d4cab29a301b
func OsmosisSwapExactAmountIn(chain *ChainNode, args...)
```

This will also further signal this new functionality only relates to Osmosis. 

I like this option better than a new Osmosis chain type.

Reply:
Good idea, added an `osmosis.go` file to the `cosmos` package with the osmosis specific functions